src/detect-bytemath.c

```diff
@@ -0,0 +1,1364 @@
+                goto error;
+                break;
+        }
+    }
+
+    switch (bmd->oper) {
+    	case DETECT_BYTEMATH_OPERATOR_LSHIFT:
+    	case DETECT_BYTEMATH_OPERATOR_RSHIFT:
+            /* nbytes has already been validated to be in [1, 10] */
+            if (bmd->nbytes > 4) {
+                SCLogError(SC_ERR_INVALID_SIGNATURE, "nbytes must be 1 through 4 (inclusive) "
+                           "when used with \"<<\" or \">>\"; %d is not valid", bmd->nbytes);
+                goto error;
+            }
+            break;
+
+        default:
+            break;
+    }
+
+    /* Set defaults for endian and base if needed */
+    if (!(bmd->flags & DETECT_BYTEMATH_FLAG_ENDIAN)) {
+        bmd->endian = DETECT_BYTEMATH_ENDIAN_DEFAULT;
+    }
+    if (!(bmd->flags & DETECT_BYTEMATH_FLAG_STRING)) {
+        bmd->base = DETECT_BYTEMATH_BASE_DEFAULT;
+    }
+
+    return bmd;
+
+ error:
+    if (bmd != NULL)
+        DetectByteMathFree(bmd);
+    return NULL;
+}
+
+/**
+ * \brief The setup function for the byte_math keyword for a signature.
+ *
+ * \param de_ctx Pointer to the detection engine context.
+ * \param s      Pointer to signature for the current Signature being parsed
+ *               from the rules.
+ * \param arg    Pointer to the string holding the keyword value.
+ *
+ * \retval  0 On success.
+ * \retval -1 On failure.
+ */
+static int DetectByteMathSetup(DetectEngineCtx *de_ctx, Signature *s, const char *arg)
+{
+    SigMatch *sm = NULL;
+    SigMatch *prev_pm = NULL;
+    DetectByteMathData *data;
+    char *rvalue = NULL;
+    int ret = -1;
+
+    data = DetectByteMathParse(arg, &rvalue);
+    if (data == NULL)
+        goto error;
+
+    int sm_list;
+    if (s->init_data->list != DETECT_SM_LIST_NOTSET) {
+        if (DetectBufferGetActiveList(de_ctx, s) == -1)
+            goto error;
+
+        sm_list = s->init_data->list;
+
+        if (data->flags & DETECT_BYTEMATH_FLAG_RELATIVE) {
+            prev_pm = DetectGetLastSMFromLists(s, DETECT_CONTENT, DETECT_PCRE, -1);
+        }
+    } else if (data->endian == DETECT_BYTEMATH_ENDIAN_DCE) {
+        if (data->flags & DETECT_BYTEMATH_FLAG_RELATIVE) {
+            prev_pm = DetectGetLastSMFromLists(s, DETECT_CONTENT, DETECT_PCRE,
+                                               DETECT_BYTETEST, DETECT_BYTEJUMP,
+                                               DETECT_BYTE_EXTRACT,
+                                               DETECT_BYTEMATH,
+                                               DETECT_ISDATAAT, -1);
+            if (prev_pm == NULL) {
+                sm_list = DETECT_SM_LIST_PMATCH;
+            } else {
+                sm_list = SigMatchListSMBelongsTo(s, prev_pm);
+                if (sm_list < 0)
+                    goto error;
+            }
+        } else {
+            sm_list = DETECT_SM_LIST_PMATCH;
+        }
+
+        if (DetectSignatureSetAppProto(s, ALPROTO_DCERPC) < 0)
```

Comment:
wonder if this is correct for a case where dcerpc is over smb. Do you have a test case for that?

Reply:
I do not have a test case ... the pattern followed is the same as that in `byte_extract`

Reply:
think we can create a SV test based on one of the smb/dce pcaps